typing/typedecl.ml

```diff
@@ -226,6 +253,34 @@ let make_constructor env type_path type_params sargs sret_type =
       widen z;
       targs, Some tret_type, args, Some ret_type
 
+(* Check that the argument to a GADT constructor is compatible with unboxing
+   the type, given the existential variables introduced by this constructor. *)
+let rec check_unboxed_gadt_arg loc ex env ty =
+  match get_unboxed_type_representation env ty with
+  | Some {desc = Tvar _ as d; _} ->
+    let f t = (Btype.repr t).desc = d in
```

Comment:
I prefer to avoid using polymorphic comparison.  What about the type_exprs `id` instead (or just using physical equality)?